### PR TITLE
build: js api

### PR DESCRIPTION
**Pull request by privatenumber**
This comes after https://github.com/oven-sh/bun/pull/2562
## Problem
JS files were essentially "path-less" (e.g. `node:path`) so they couldn't import relative paths, and thus couldn't reuse code.

## Changes
- Moves `src/bun.js/*.exports.js` files to `src/bun.js/src/*.exports.js`
- Build using esbuild to `src/bun.js/dist/*.exports.js`

## Other info
- In the near future, we want to build these using bun

- In the future, I wonder if Node.js APIs can be aliases (e.g. `node:path`), so their real FS path allows them to be used to resolve relative paths. This way, we can skip the build step during development. @Jarred-Sumner

### CODE REVIEW

File: src/bun.js/javascript.zig [hunk 1]
```diff
@@ -433,7 +433,7 @@ pub const VirtualMachine = struct {
     is_us_loop_entered: bool = false,
     pending_internal_promise: *JSC.JSInternalPromise = undefined,
     auto_install_dependencies: bool = false,
-    load_builtins_from_path: []const u8 = "",
+    load_builtins_from_path: []const u8 = "dist",
```
Comment (privatenumber) [inline src/bun.js/javascript.zig:436]: Is this correct?

File: package.json [hunk 2]
```diff
@@ -11,6 +12,7 @@
   },
   "private": true,
   "scripts": {
+    "build-js-api": "esbuild --bundle src/bun.js/src/*.exports.js --outdir=src/bun.js/dist/ --format=esm --platform=node",
```
Comment (privatenumber) [inline package.json:15]: Should I add a watch version: `build-js-api:dev`?

Is `build-js-api` too wordy?

File: Makefile [hunk 3]
```diff
@@ -814,13 +814,15 @@ runtime_js:
...
 	@NODE_ENV=development esbuild --define:process.env.NODE_ENV="development" --target=esnext  --bundle src/runtime/index-with-refresh.ts --format=iife --platform=browser --global-name=BUN_RUNTIME --external:/bun:* > src/runtime.out.refresh.js; cat src/runtime.footer.with-refresh.js >> src/runtime.out.refresh.js
 	@NODE_ENV=development esbuild --define:process.env.NODE_ENV="development" --target=esnext  --bundle src/runtime/index-without-hmr.ts --format=iife --platform=node --global-name=BUN_RUNTIME --external:/bun:* > src/runtime.node.pre.out.js; cat src/runtime.node.pre.out.js src/runtime.footer.node.js > src/runtime.node.out.js
 	@NODE_ENV=development esbuild --define:process.env.NODE_ENV="development" --target=esnext  --bundle src/runtime/index-without-hmr.ts --format=iife --platform=node --global-name=BUN_RUNTIME --external:/bun:* > src/runtime.bun.pre.out.js; cat src/runtime.bun.pre.out.js src/runtime.footer.bun.js > src/runtime.bun.out.js
+	@NODE_ENV=development esbuild --bundle src/bun.js/src/*.exports.js --outdir=src/bun.js/dist/ --format=esm --platform=node
```
Comment (privatenumber) [inline Makefile:825]: The `NODE_ENV` isn't actually necessary here but I added it to follow the convention. But personally would prefer to keep it minimal. LMK if I can remove.

File: src/bun.js/javascript.zig (hunk 1 above)
Comment (Jarred-Sumner) [reply src/bun.js/javascript.zig:436]: No, you’ll need to modify the usages of embedFile in module_loader.zig

**Comment by privatenumber** [comment]
Hopefully we won't have to move forward with this PR.

@Jarred-Sumner is considering adding an alias for internal utils, so bundling is not necessary to share code.

**Comment by xhyrom** [comment]
i think its weird to have src/ in src/, something like packages/* would be better but it requires to refactor whole repo but I'm not sure if zig supports monorepo

File: package.json (hunk 2 above)
Comment (Electroid) [reply package.json:15]: Maybe just `build:js`?

File: Makefile (hunk 3 above)
Comment (Electroid) [reply Makefile:825]: If there isn't a material difference in what esbuild outputs, seems fine to remove.

**Comment by Jarred-Sumner** [comment]
what if we did a toplevel `lib` or `runtime` directory? The existence of `src/bun.js` is strange too

**Comment by privatenumber** [comment]
`lib` would be nice to follow Node.js. Could it be ambiguous in the context of Bun?

`runtime` makes sense to me but I'm afraid about introducing another way to describe it. Node.js also uses ["core modules"](https://nodejs.org/api/modules.html#core-modules) or ["builtin modules"](https://nodejs.org/api/module.html#modulebuiltinmodules).

In any case, I'll likely scrap this PR and consider the rename for a follow-up PR related to the internals/alias refactor.

**Comment by Electroid** [comment]
Thanks for working on this, we recently improved the JS files in the repo.
